src/plugins/workspace/server/permission_control/client.ts

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { i18n } from '@osd/i18n';
+import {
+  ACL,
+  SavedObjectsBulkGetObject,
+  SavedObjectsServiceStart,
+  Logger,
+  OpenSearchDashboardsRequest,
+  Principals,
+  SavedObject,
+  WORKSPACE_TYPE,
+  Permissions,
+  HttpAuth,
+} from '../../../../core/server';
+import { WORKSPACE_SAVED_OBJECTS_CLIENT_WRAPPER_ID } from '../../common/constants';
+import { getPrincipalsFromRequest } from '../utils';
+
+export type SavedObjectsPermissionControlContract = Pick<
+  SavedObjectsPermissionControl,
+  keyof SavedObjectsPermissionControl
+>;
+
+export type SavedObjectsPermissionModes = string[];
+
+export class SavedObjectsPermissionControl {
+  private readonly logger: Logger;
+  private _getScopedClient?: SavedObjectsServiceStart['getScopedClient'];
+  private auth?: HttpAuth;
+  /**
+   * Returns a saved objects client that is able to:
+   * 1. Read objects whose type is `workspace` because workspace is a hidden type and the permission control client will need to get the metadata of a specific workspace to do the permission check.
+   * 2. Bypass saved objects permission control wrapper because the permission control client is a dependency of the wrapper to provide the ACL validation capability. It will run into infinite loop if not bypass.
+   * @param request
+   * @returns SavedObjectsContract
+   */
+  private getScopedClient(request: OpenSearchDashboardsRequest) {
+    return this._getScopedClient?.(request, {
+      excludedWrappers: [WORKSPACE_SAVED_OBJECTS_CLIENT_WRAPPER_ID],
+      includedHiddenTypes: [WORKSPACE_TYPE],
+    });
+  }
+
+  constructor(logger: Logger) {
+    this.logger = logger;
+  }
+
+  private async bulkGetSavedObjects(
```

Comment:
When encountering a large number of SavedObjects cases, it's worth considering whether to reduce the number of DDB queries. One approach could involve iterating over the SavedObjects array, grouping them by type, and creating batches of requests. This may help in reducing the overall number of DDB queries.

Reply:
Thanks for your suggestions. For DDB queries, we can add bulk get optimize logic in the future version if grouping by type will more efficient.